### DISCUSSION TITLE
PJFNK vs Newton ADNodalBC

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#23931] [Q&A General] [answered]
Background, I implemented an embedded Dirichlet BC method using the ADNodalBC class. When I run the test using the Newton Method, it can take ~100 iterations to converge. When I run it using PJFNK, it takes 5 to converge. The solutions are the same in both cases, but somehow, PJFNK gets there with fewer iterations. Is there a difference in how the two implement the BC for the Jacobain? I want to run with NEWTON because the final use case will solve NS equations and I had better performance with NEWTON for NS.

**Comment by GiudGiud** [comment, marked as answer]
Jacobian's probably not exact. 
what does this return?
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by maxnezdyur** [reply]
Not sure if it will work in my case (from the link)`analyzejacobian.py turns off boundary all conditions, e.g. it currently tests kernels only and not integrated boundary conditions.`. I am not using any kernel that is not a MOOSE kernel. I just have a different BC class.

**Reply by GiudGiud** [reply]
Ah good point. 

Can you share the source for this BC?

**Reply by lindsayad** [reply]
Something definitely is not right. Seeing the source (and input file) would be great. When debugging Jacobians, the petsc command line options `-snes_test_jacobian -snes_test_jacobian_view` are great when you have relatively few degrees of freedom.

**Reply by lindsayad** [reply]
you're not doing any AD in your `compute_boundary_value` routine. You're implicitly converting a `Real` to an `ADReal` in your return. I'll work on a commit for you that adds the AD

**Reply by maxnezdyur** [reply]
Thanks for looking into that. Before, I did have it all AD, but with how I got the Shape Functions at a point, I wasn't sure if it mattered. Hopefully, that gets Newton converging faster. My main concern is that PJFNK was still converging rapidly, and Newton was not. If I just threw away the AD terms, then I would expect both to have the same bad convergence.

**Reply by lindsayad** [reply]
I got one nonlinear iteration after my fixes. This also requires a change in `ADNodalBC` which I'll make a PR for soon

**Reply by maxnezdyur** [reply]
Wow!

**Reply by lindsayad** [reply]
see https://github.com/maxnezdyur/insect/pull/1 and https://github.com/idaholab/moose/pull/23941
